core/src/unprocessed_packet_batches.rs

```diff
@@ -1,209 +1,353 @@
 use {
-    retain_mut::RetainMut,
+    min_max_heap::MinMaxHeap,
     solana_perf::packet::{limited_deserialize, Packet, PacketBatch},
+    solana_runtime::bank::Bank,
     solana_sdk::{
-        hash::Hash, message::Message, short_vec::decode_shortu16_len, signature::Signature,
-        transaction::VersionedTransaction,
-    },
-    std::{
-        collections::{HashMap, VecDeque},
-        mem::size_of,
+        hash::Hash,
+        message::{Message, VersionedMessage},
+        short_vec::decode_shortu16_len,
+        signature::Signature,
+        transaction::{Transaction, VersionedTransaction},
     },
+    std::{cmp::Ordering, collections::HashMap, mem::size_of, rc::Rc, sync::Arc},
 };
 
+#[derive(Debug, Default, PartialEq, Eq)]
+struct ImmutableDeserializedPacket {
+    original_packet: Packet,
+    versioned_transaction: VersionedTransaction,
+    message_hash: Hash,
+    is_simple_vote: bool,
+    fee_per_cu: u64,
+}
+
 /// Holds deserialized messages, as well as computed message_hash and other things needed to create
 /// SanitizedTransaction
-#[derive(Debug, Default)]
+#[derive(Debug, Default, Clone, PartialEq, Eq)]
 pub struct DeserializedPacket {
-    pub versioned_transaction: VersionedTransaction,
-    pub message_hash: Hash,
-    pub is_simple_vote: bool,
-}
-
-/// Defines the type of entry in `UnprocessedPacketBatches`, it holds original packet_batch
-/// for forwarding, as well as `forwarded` flag;
-/// Each packet in packet_batch are deserialized upon receiving, the result are stored in
-/// `DeserializedPacket` in the same order as packets in `packet_batch`.
-#[derive(Debug, Default)]
-pub struct DeserializedPacketBatch {
-    pub packet_batch: PacketBatch,
+    immutable_section: Rc<ImmutableDeserializedPacket>,
     pub forwarded: bool,
-    // indexes of valid packets in batch, and their corresponding deserialized_packet
-    pub unprocessed_packets: HashMap<usize, DeserializedPacket>,
 }
 
-/// References to a packet in `UnprocessedPacketBatches`, where
-/// - batch_index references to `DeserializedPacketBatch`,
-/// - packet_index references to `packet` within `DeserializedPacketBatch.packet_batch`
-#[derive(Debug, Default)]
-pub struct PacketLocator {
-    #[allow(dead_code)]
-    batch_index: usize,
-    #[allow(dead_code)]
-    packet_index: usize,
+impl DeserializedPacket {
+    pub fn new(packet: Packet, bank: &Option<Arc<Bank>>) -> Option<Self> {
```

Comment:
```suggestion
    pub fn new(packet: Packet, bank: Option<&Arc<Bank>>) -> Option<Self> {
```

`&None` is weird looking

Reply:
fixed up here: https://github.com/solana-labs/solana/pull/24187/commits/7a2794ea62d31f26ab6f6492a63956eaab40937f